File: Python DS & Algo - Runestone Academy/7. Trees/C++/binary_heap.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

class BinaryHeap{
private:
    vector <int> heap {0};
    int size = 0;
public:
    void insert(int item){
        heap.push_back(item);
        size++;
        perc_up(size);
    }
    void perc_up(int idx){
         while(idx / 2 > 0){
             int parent_idx = idx / 2;
             if(heap[idx] < heap[parent_idx]){
                 swap(heap[idx], heap[parent_idx]);
                 idx /= 2;
             }
             else break;
         }
    }
    int del_min(){
        int return_num = heap[1];
        heap[1] = heap[size];
        size--;
        heap.pop_back();
        perc_down(1);
        return return_num;
    }
    void perc_down(int idx){
        while(idx * 2 <= size){
            int min_child_idx = find_min_child_idx(idx);
            if(heap[idx] > heap[min_child_idx])
                swap(heap[idx], heap[min_child_idx]);
            idx = min_child_idx;
        }
    }
    int find_min_child_idx(int idx){
        // If no right child, return left child index
        int return_idx;
        if(idx * 2 + 1 > size)
            return_idx = idx * 2;
        else{
            if(heap[idx * 2] < heap[idx * 2 + 1])
                return_idx = idx * 2;
            else
                return_idx = idx * 2 + 1;
        }
        return return_idx;
    }
    void build_heap(vector <int> arr){
        size = arr.size();
        for(int i = 0; i < size; i++)
            heap.push_back(arr[i]);
        int idx = size / 2;
        while(idx > 0)
            perc_down(idx--);
    }
};


int main(void){
    srand(time(NULL));
    int size = 10;
    vector <int> arr;
    BinaryHeap bh;

    // If doing multiple insertion (O (n log n)), use below for loop.
    // for(int i = 0; i < size; i++){
    //     int rand_num = rand() % 100;
    //     arr.push_back(rand_num);
    //     bh.insert(rand_num);
    // }

    for(int i = 0; i < size; i++)
        arr.push_back(rand() % 100);

    bh.build_heap(arr);

    sort(arr.begin(), arr.end());
    for(int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;

    for(int i = 0; i < size; i++)
        cout << bh.del_min() << " ";
    cout << endl;
}

File: Python DS & Algo - Runestone Academy/7. Trees/binary_heap.py
import random

class BinaryHeap:
    def __init__(self):
        self.heap = [0]
        self.size = 0

    def insert(self, item):
        self.heap.append(item)
        self.size += 1
        self.perc_up(self.size)


    def perc_up(self, idx):
        while idx // 2 > 0:
            if self.heap[idx] < self.heap[idx // 2]:
                self.heap[idx], self.heap[idx // 2] = self.heap[idx//2], self.heap[idx]
            idx //= 2

    def del_min(self):
        return_val = self.heap[1]
        self.heap[1] = self.heap[self.size] # replacing root node with last node in heap
        self.heap.pop()
        self.size -= 1
        self.perc_down(1)
        return return_val

    def perc_down(self, idx):
        while idx * 2 <= self.size:
            min_child_idx = self.get_min_child_idx(idx)
            if self.heap[idx] > self.heap[min_child_idx]:
                self.heap[idx], self.heap[min_child_idx] = self.heap[min_child_idx], self.heap[idx]
            idx = min_child_idx

    def get_min_child_idx(self, idx):
        if (idx * 2 + 1) > self.size:
            # If there is no right child, return left child's index
            return idx * 2
        else:
            if self.heap[idx * 2] < self.heap[idx * 2 + 1]:
                return idx * 2
            else:
                return idx * 2 + 1

    def build_heap(self, lst):
        self.size = len(lst)
        self.heap = [0] + lst[:]
        idx = len(lst) // 2
        while idx > 0:
            self.perc_down(idx)
            idx -= 1

    def __str__(self):
        return ' '.join(str(num) for num in self.heap)

size = 10
lst = [random.randint(1, 100) for i in range(size)]
bh = BinaryHeap()
bh.build_heap(lst)
print('Random List:\t\t\t', lst)
print('Sorted List:\t\t\t', sorted(lst))
print('Binary Heap Deleted Minimums:\t', [bh.del_min() for i in range(size)])
